### PR TITLE
LINKTA-94: Back-end sends information to front-end

### PR DESCRIPTION
**Pull request by dcpavel**
## Description

This creates a server.ts file in the server directory and has a structure for handling errors from middleware.

Added nodemon to the dependencies to aid in development.

Fixes # (issue) [LINKTA-94](https://linkta.atlassian.net/browse/LINKTA-94)

## Type of change

- [ ]  🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ✨  New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 Documentation update
- [ ] Other 🗒️ ______________

## How Has This Been Tested?

- [x] Manual testing to make sure that the server starts using nodemon

## Checklist:

- [x] I have performed a self-review of my own code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [x] I have reviewed the [**Security Best Practices Document**](SECURITY_BEST_PRACTICES.md).
- [x] I have followed the [**Naming Conventions Guide**](NAMING_CONVENTIONS_GUIDE.md) for my changes.
- [x] I have ensured that my pull request title is descriptive.

## Additional Information:

[LINKTA-94]: https://linkta.atlassian.net/browse/LINKTA-94?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: server/server.ts
Comment (LucieSeidler) [inline server/server.ts:1]: Doesn't there need to be global router that loads the page?

File: server/utils/base.ts
Comment (LucieSeidler) [inline server/utils/base.ts:1]: I'm not sure the file name is sufficiently descriptive

File: server/types.ts
Comment (LucieSeidler) [inline server/types.ts:1]: Will this hold all types for the backend? I'd suggest file naming improvements

**Review by LucieSeidler — CHANGES_REQUESTED** [review]

File: server/server.ts
Comment (dcpavel) [reply server/server.ts:1]: That is the first app.get() called. We don't have a frontend yet, so this will need to change once we do.

**Comment by dcpavel** [comment]
I've changed some naming and locations. What do you think of the new structure and naming?

File: server/server.ts
Comment (dcpavel) [reply server/server.ts:1]: The first app.get('/', ...) loads the page. I have a PORT variable of 8000, and I get the message when navigating to http://localhost:8000

**Review by LucieSeidler — APPROVED** [review]

**Review by kinskiwu — COMMENTED** [review]
Fantastic effort on establishing the back-end infrastructure! To further bolster our setup, I recommend the following enhancements:

1. **Consolidate Type Definitions**
   - Introduce an `index.ts` in the `types` folder to streamline imports:
     ```typescript
     // types/index.ts
     export * from './middleware.d.ts';
     ```

2. **Organize Middleware**
   - Shift the global error handler and future middleware to a `middleware` directory for better organization:
     ```
     /middleware
       - globalErrorHandler.ts
     ```

3. **Enhance Server Port Config**
   - Ensure `process.env.PORT` is always defined by setting a default value:
     ```typescript
     const PORT = process.env.PORT || '3000';
     ```

4. **Refactor Server Initialization**
   - Encapsulate initialization in a `startApp` function, making the process clearer and more maintainable. This setup also allows for pre-listening async operations and allow for future incorporation with database connection.
   - Add SIGINT and SIGTERM to allow the server to shut down more gracefully (optional)
   - Integrate a check to skip server startup during test runs to avoid clashes with Jest:
     ```typescript
     async function startApp() {
       try {
         const app = express();
         const PORT = process.env.PORT || '3000';

         // Add database connect here

         const server = app.listen(PORT, () => {
           console.log(`Server running on http://localhost:${PORT} in ${process.env.NODE_ENV} mode`);
         });

         const shutdownApp = () => {
           server.close(() => {
             console.log('Server closed');
             // Add database disconnect here
             process.exit(0);
           });
         };

         // Listen for termination signals to gracefully shut down the server
         process.on('SIGTERM', shutdownApp);
         process.on('SIGINT', shutdownApp);
       } catch (error) {
         console.error(`App startup failure: ${error}`);
         process.exit(1);
       }
     }

     if (process.env.NODE_ENV !== 'test') {
       startApp();
     }
     ```

**Comment by dcpavel** [comment]
As Lucie mentioned, fantastic notes! I have a couple of questions regarding your suggestions:

1. Why is an index.ts exporting everything better than exporting from individual files? As this application grows, there will be other type definition files. In this suggestion, we will always import the needed type from index.ts, instead of the particular type definition file. Doesn't this obfuscate the link between type and it's use?

2. I'm not exactly using the createError function as middleware, but I wasn't sure what a good name was. I do like the idea of moving the global error handler to a middleware folder, and then using it in server.ts, do you have a suggestion for what todo with the createError function that is actually meant to be used with the controllers?

3. This is a good suggestion. Do we have a preferred port to use for the backend? I believe Vite defaults to 5173, so I think 3000 should be safe.

4. I love this change, but I do think we will want to start up the application for integration and end-to-end testing. Any thoughts @LucieSeidler @ChTiSh @AnatoliySokolov98?

**Comment by kinskiwu** [comment]
Thanks for the feedback! 

Let me address Q1: using an index.ts (or index.d.ts - depending on our needs) file for exporting everything centralizes and simplifies the management of exports, especially as an application grows. This method doesn't obfuscate the link between a type and its use but rather organizes it in a way that makes imports cleaner and reduces the complexity of import paths across the project. Here are the key benefits:

1. **Simplified Imports**: Centralizing exports in index.ts reduces complex relative paths, making imports cleaner.
2. **Ease of Refactoring**: Changes to file structures only require updates in index.ts, not everywhere a type is imported.
3. **Enhanced Discoverability**: An index.ts provides a quick overview of all available types, improving clarity for developers.
4. **Fewer Merge Conflicts**: Centralizing exports can lead to fewer conflicts during collaborative development.

Resources: 
- https://stackoverflow.com/questions/51516394/what-is-the-main-usage-of-index-d-ts-in-typescript
- https://medium.com/@bobjunior542/maximize-typescript-productivity-use-the-index-file-now-cac2f4f4d1ed#:~:text=The%20index%20file%20can%20be,from%20a%20single%20index%20file.

File: server/middleware/errorHandling.ts
Comment (kinskiwu) [inline server/middleware/errorHandling.ts:1]: Just curious, for the globalErrorHandler, is `_: Request` intentional? 

I think it is fine to keep these two together in this file. I normally use a service folder to keep non-middleware or helper functions but I can see there is a benefit of keeping these two together.

**Review by kinskiwu — APPROVED** [review]
Looks good after addressing `_: Request`. Thanks for the great work^^

File: server/middleware/errorHandling.ts
Comment (dcpavel) [reply server/middleware/errorHandling.ts:1]: Yes, you are going to see a lot of this in middleware. We have a warning on unused variables, but you need to pass on the request. If I called it req, we would get the linting error.

Comment (kinskiwu) [reply server/middleware/errorHandling.ts:1]: Thanks for the clarification! I learned something new today😊

**Pull request merged**